File: shared/src/main/java/ru/privatenull/pnauth/platform/TaskRegistry.java
package ru.privatenull.pnauth.platform;

import java.time.Duration;
import java.util.Collection;
import java.util.Optional;
import java.util.UUID;

/** Named task directory. Replacing a task atomically cancels its previous instance. */
public interface TaskRegistry {
    TaskHandle delayed(String owner, String taskId, Duration delay, Runnable task);
    TaskHandle delayed(String owner, String taskId, PnPlayer player, Duration delay, Runnable task);
    TaskHandle repeating(String owner, String taskId, Duration initialDelay, Duration interval, Runnable task);
    TaskHandle repeating(String owner, String taskId, PnPlayer player, Duration initialDelay,
                         Duration interval, Runnable task);
    Optional<TaskHandle> find(String owner, String taskId, UUID playerId);
    Collection<TaskHandle> ownedBy(String owner);
    boolean cancel(String owner, String taskId, UUID playerId);
    int cancelAll(String owner);
    int cancelAll(UUID playerId);
    int cancelAll();
}

File: shared/src/main/kotlin/ru/privatenull/pnauth/platform/TaskRegistry.kt
package ru.privatenull.pnauth.platform

import java.time.Duration
import java.util.Optional
import java.util.UUID

/** Named task directory. Replacing a task atomically cancels its previous instance. */
interface TaskRegistry {
    fun delayed(owner: String, taskId: String, delay: Duration, task: Runnable): TaskHandle
    fun delayed(owner: String, taskId: String, player: PnPlayer, delay: Duration, task: Runnable): TaskHandle
    fun repeating(owner: String, taskId: String, initialDelay: Duration, interval: Duration, task: Runnable): TaskHandle
    fun repeating(
        owner: String,
        taskId: String,
        player: PnPlayer,
        initialDelay: Duration,
        interval: Duration,
        task: Runnable
    ): TaskHandle

    fun find(owner: String, taskId: String, playerId: UUID?): Optional<TaskHandle>
    fun ownedBy(owner: String): Collection<TaskHandle>
    fun cancel(owner: String, taskId: String, playerId: UUID?): Boolean
    fun cancelAll(owner: String): Int
    fun cancelAll(playerId: UUID): Int
    fun cancelAll(): Int
}
